earth2mip/lagged_ensembles/core.py

```diff
@@ -53,69 +56,50 @@ async def yield_lagged_ensembles(
     finished = set()
     ensemble = {}
 
-    obs_buffer = deque([])
-
-    for i in range(n + world_size):
-        obs_buffer.append(await observations[i])
+    n_iter = nt // world_size
+    if nt % world_size != 0:
+        n_iter += 1
 
-    n_iter = int(nt // world_size)
-    assert nt % world_size == 0  # noqa
-
-    buffers = None
+    recv_buffer = None
 
     for i0 in range(n_iter):
-        for k in range(world_size):
-            i = world_size * i0 + k
-            if i + n + 1 < nt:
-                obs_buffer.append(await observations[i + n + 1])
-
         i = world_size * i0 + rank
         nsteps = min(nt - world_size * i0 - 1, n)
 
         lead_time = -1
-        async for y in forecast[i]:
+
+        single_forecast = forecast[i] if i < nt else repeat(None)
```

Comment:
Why not just `yield None` instead this repeat. The present make the iterator infinite when this isn't true. Is that intended?

Reply:
All the ranks need to yield the same number of steps (n) or they will deadlock when the communication happens.

Reply:
Ah, got it. Thanks!